gracefully handle apps without `package.json` files (fixes #50)

cvan opened the pull request:
@mykmelez: could use some help on the failing tests. (I typically use different test runners, so the whole `SIGINT`, `stdout`/`stderror`, streams, Promise chain is a bit unfamiliar to me.)

cvan committed:
gracefully run/package apps with missing/invalid `package.json` files (fixes #50)

mykmelez committed:
return transformed result; log warnings to stdout; clarify stderr failure

Ensure the transform function in readProjectMetadata returns the result,
since the function that calls it expects to use its return value.

Log expected warnings to stdout rather than stderr, since the tests
automatically fail on output to stderr.  (Alternately, we could make the tests
expect those warnings.)

When tests fail because of output to stderr, make the cause of the failure
more obvious, logging the message 'test failure due to stderr output'.

mykmelez committed:
appPackageJson.pkg -> appPackageJson

appPackageJson is appPackageResult.pkg, and it doesn't itself have a pkg
property, so these two references to appPackageJson.pkg undoubtedly meant
to refer to appPackageJson itself.

mykmelez committed:
stop rejecting tests due to stderr output

mykmelez committed:
remove run-url-bad-package-name.js as it doesn't test a bad package name

When you run a URL, qbrt uses the shell package to run it, and it has a good
package name.  So this test doesn't actually do what it's name suggests.
Nor is it necessary to test this, as run-path-bad-package.name.js is sufficient
to confirm that a bad package name will trigger a failure.

mykmelez committed:
ensure result.pkg and result.path are defined; avoid finding qbrt's package.json

Ensure result.pkg and result.path are defined if we don't find a package.json
file, since the code assumes those properties are defined.

Avoid finding qbrt's package.json when traversing the directory tree
for the package.json file belonging to the test app that is missing that file.
This ensures that we actually test what happens when an app is missing
its package.json file, rather than testing what happeens when an app reuses
qbrt's package.json file (which, by the way, results in failure).

mykmelez commented:
> @mykmelez: could use some help on the failing tests. (I typically use different test runners, so the whole SIGINT, stdout/stderror, streams, Promise chain is a bit unfamiliar to me.)

I've submitted https://github.com/cvan/qbrt/pull/1 to fix the test failures. I'm not entirely sure about some of the fixes—I may want to revisit them and do something else instead—but that PR will at least unblock this PR.

mykmelez committed:
fix temp dir path on Linux

mykmelez committed:
let qbrt die on bad package name

cvan commented:
many thanks, much appreciated. will address feedback and set up Travis CI on my fork. I agree re: the warnings.

cvan committed:
Merge pull request #1 from mykmelez/cvan-pkg-json-optional

fix test failures on pkg-json-optional branch

cvan committed:
remove "Reading package …" messages

cvan committed:
per feedback, remove `package.json` warnings

cvan commented:
@mykmelez: for some reason [Taskcluster](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FXd4JYquTTx6U_eooYPkoZQ%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=&highlightEnd=&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false) is failing on `test/run-path-missing-package.js` on Linux, but it seems to be fine on Windows and on Travis CI

cvan committed:
remove unused variable

cvan committed:
remove `cli` npm scripts; remove unused `read-pkg` dependency

cvan committed:
add back `start` npm script

cvan committed:
add back `start` npm script

mykmelez commented:
> @mykmelez: for some reason Taskcluster is failing on test/run-path-missing-package.js on Linux, but it seems to be fine on Windows and on Travis CI (and locally)

This error message in the log identifies the failure:

> (node:1885) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): TypeError: Path must be a string. Received { qbrt: './bin/cli.js' }

When an app is missing a package.json manifest, qbrt now traverses its directory tree looking for one. And if that app is in a subdirectory of the qbrt directory (f.e. test/hello-world-missing-package/), then it find's qbrt's package.json manifest.

And that manifest specifies `"bin": "./bin/cli.js"`, which [normalize-package-data](https://www.npmjs.com/package/normalize-package-data) converts into something like `"bin": { qbrt: './bin/cli.js' }`, per this rule:

> If bin field is a string, then bin field will become an object with name set to the value of the name field, and bin set to the original string value.

Meanwhile, cli.js sets *appPackageResult.pkg.main* to `appPackageResult.pkg.main || appPackageResult.pkg.bin || 'index.js'`; and since qbrt's manifest doesn't have a *main* property, cli.js sets it to the *bin* property, which is the aforementioned object, triggering the error.

I tried to work around this issue in https://github.com/mozilla/qbrt/pull/83/commits/283bbc934010248854333f4496e892789a2b1698 by copying test/hello-world-missing-package/ to the temporary directory and then running it from there. But the temporary directory appears to be `/repo` in the TaskCluster Docker image, which is presumably also the directory where it clones qbrt, thus foiling my ingenious workaround. ;-)

So I'm not sure what to do next. One option would be to find some other directory where we can copy the app on Linux (at least in TaskCluster, but ideally we'd have a general solution). Alternately, we could check *appPackageResult.pkg.bin* to ensure it's a reasonable value before trying to use it. Or we could remove the read-pkg-up functionality, although presumably it's useful, so that's a last resort.

mykmelez commented:
> So I'm not sure what to do next. One option would be to find some other directory where we can copy the app on Linux (at least in TaskCluster, but ideally we'd have a general solution). Alternately, we could check appPackageResult.pkg.bin to ensure it's a reasonable value before trying to use it. Or we could remove the read-pkg-up functionality, although presumably it's useful, so that's a last resort.

[Further reading](https://stackoverflow.com/questions/38747082/docker-container-has-no-tmpdir-environmental-variable) suggests that this is probably because the Docker image has no TMPDIR environment variable, which we can fix with an ENV step in the Dockerfile. However, currently we're using the stock node:6 Docker image, and I'd like to avoid building a custom image if possible.

I've tried a couple of other ways of setting TMPDIR over in #84, but so far no luck.

mykmelez committed:
run tests as non-root user

npm sets TMPDIR to the cwd if running as the root user, per https://github.com/npm/npm/issues/4531, which is the, ahem, root cause of the test failure in run-path-missage-package.js.  There are two workarounds: invoke npm with --unsafe-perm or run it as a different user.  This commit attempts to do the latter.

mykmelez committed:
Merge branch 'pkg-json-optional' of https://github.com/cvan/qbrt into cvan-pkg-json-optional

mykmelez committed:
remove debugging code

mykmelez commented:
@cvan Figured out and fixed in https://github.com/cvan/qbrt/pull/2.

cvan commented:
Thanks for the explanation. I completely missed that; that totally makes sense. Looking at PR.

mykmelez commented on test/package-main-path.js:
```diff
@@ -55,8 +55,9 @@ new Promise((resolve, reject) => {
+    // console.error('test failure due to stderr output');
     console.error(error);
-    reject(error);
+    // reject(error);
```
Now that you've removed the "reading package" messages and "missing package.json" warnings, it should be possible to re-enable this error message and test failure on output to stderr.

mykmelez replied:
(You should now be able to do this in the two other places where I disabled that error message and test failure.)

mykmelez replied:
I've made these changes in 556b8a9b749e04439ec6f3224d2f982b1f60a94a.

mykmelez commented on test/run-path-bad-package-name.js:
```diff
@@ -0,0 +1,70 @@
+    // if (outputRegex.test(totalError) && !quitting) {
+    //   child.kill('SIGINT');
+    //   quitting = true;
+    // }
```
I guess this commented-out code could just be removed, as it isn't needed.

mykmelez replied:
I've made this change in 834ec3c1415916b9875d1c1f19c92752000bbeea. I also made a minor change to this function in a82416ed0e9ba54997721e8dfa853e469df3132e.

cvan committed:
Merge pull request #2 from mykmelez/cvan-pkg-json-optional

fix run-path-missing-package.js on TaskCluster

mykmelez commented:
I've fixed up the last few nits, so I'll merge this with those changes.

cvan commented:
thanks for making these changes. sorry, I've been away working on a different project for a bit. your help's much appreciated 👍
